packages/editor/src/components/use-global-styles-data/index.js

```diff
@@ -0,0 +1,79 @@
+/**
+ * External dependencies
+ */
+import deepmerge from 'deepmerge';
+import { isPlainObject } from 'is-plain-object';
+
+/**
+ * WordPress dependencies
+ */
+import { useSelect } from '@wordpress/data';
+import { useMemo } from '@wordpress/element';
+import { store as coreStore } from '@wordpress/core-data';
+
+const DEFAULT_STYLES = {};
+
+function useGlobalStylesBaseData() {
+	const baseConfig = useSelect( ( select ) => {
+		return select(
+			coreStore
+		).__experimentalGetCurrentThemeBaseGlobalStyles();
+	}, [] );
+
+	return [ !! baseConfig, baseConfig?.styles ?? DEFAULT_STYLES ];
+}
+
+function useGlobalStylesUserData() {
+	return useSelect( ( select ) => {
+		const {
+			getEditedEntityRecord,
+			hasFinishedResolution,
+			__experimentalGetCurrentGlobalStylesId,
+		} = select( coreStore );
+
+		const globalStylesId = __experimentalGetCurrentGlobalStylesId();
+		const record = globalStylesId
+			? getEditedEntityRecord( 'root', 'globalStyles', globalStylesId )
+			: undefined;
+
+		let hasResolved = false;
+		if (
+			hasFinishedResolution( '__experimentalGetCurrentGlobalStylesId' )
+		) {
+			hasResolved = globalStylesId
+				? hasFinishedResolution( 'getEditedEntityRecord', [
+						'root',
+						'globalStyles',
+						globalStylesId,
+				  ] )
+				: true;
+		}
+
+		return [ hasResolved, record?.styles ?? DEFAULT_STYLES ];
+	}, [] );
+}
+
+function mergeBaseAndUserStyles( base, user ) {
+	return deepmerge( base, user, {
+		// We only pass as arrays the presets,
+		// in which case we want the new array of values
+		// to override the old array (no merging).
+		isMergeableObject: isPlainObject,
+	} );
+}
+
+export function useGlobalStylesData() {
+	const [ isBaseStylesReady, baseStyles ] = useGlobalStylesBaseData();
+	const [ isUserStylesReady, userStyles ] = useGlobalStylesUserData();
+	const mergedStyles = useMemo(
+		() => mergeBaseAndUserStyles( baseStyles, userStyles ),
+		[ baseStyles, userStyles ]
+	);
+
+	return {
+		isReady: isBaseStylesReady && isUserStylesReady,
+		base: baseStyles ?? DEFAULT_STYLES,
+		user: userStyles ?? DEFAULT_STYLES,
```

Comment:
Is it a good idea to return base, user and merged, or should we just return merged?

Reply:
I did wonder about this myself.

It could be argued returning the different origins' data is more future-proof. For example, the style inheritance UI could differ based on at which level the style was modified. 

The current approach also reflects the data in the site editor's global styles context providing more consistency. Which could help consolidate some of this down the road.

Reply:
I'm not entirely sure here. If one think about the block editor package as a standalone package, it doesn't really know that there are multiple levels of style configurations and probably doesn't matter for blocks where the style is coming from. I'd love to have more details about the potential UIs before committing to this.

Reply:
> think about the block editor package as a standalone package, it doesn't really know that there are multiple levels of style configurations

That's a good point, thanks @youknowriad 👍 

One other option I was pondering was only adding the `merged` property for now keeping the door open to add the others later if needed.

To your point though maybe I should switch this PR to only return the merged values.

@richtabor can probably confirm the style inheritance UI work only needs the merged data.

Reply:
>  I'd love to have more details about the potential UIs before committing to this.

I don't think that we need much (perhaps any) UI for style inheritance. If we can get the default values available on blocks, and leverage the existing reset/clear UI. 

For example, this could be an unedited group block, with its styles derived from global styles, with a font size of XL applied locally: 

![CleanShot 2024-05-09 at 17 02 59](https://github.com/WordPress/gutenberg/assets/1813435/8bbc2621-1951-4eac-8225-c2504949ca5a)

I'd expect resetting the font size would set it to whatever the current global styles value is (regardless of if its `theme` or `user`).

Reply:
So it looks like we only have a use case for the merged data so far. I'll get this updated to only add the merged styles to the block editor settings.

Reply:
I've updated this PR to return only the merged styles data. Let me know if there is anything else that needs tweaking.

Reply:
Rethinking this a bit, I'd like to reduce some code duplication with the site editor's global styles context setup. 

This would require the user, base, and merged configs to be exported from a new hook in the editor package. The editor setting however will still only be the merged style data as discussed.

I'll update again when this is ready for another look.

Reply:
There's an alternate approach to this PR up in https://github.com/WordPress/gutenberg/pull/61556.

TL;DR

The new approach relocates the collection of GlobalStylesContext data to the editor package. This way we have a single location that collects and merges the global styles data. So when block style variations need to be resolved within that merged data, it only happens once and is available in both editors.

After moving the GlobalStylesContext it seemed to make sense that the GlobalStylesProvider was also in the editor package. Happy to move if required.